### DISCUSSION TITLE
assistance request: SynoAI utilizing CodeProjectAI

### DISCUSSION THREAD

**Discussion by JoeyJoe6** [djdd87/SynoAI#179] [Q&A]
I am attempting to utilize synoAI with codeprojectAI (GPU version) and having some small issues. Below is some background info on my setup, another set of eyes to comment would be greatly appreciated. This is a new install i am testing using a DVA1622. I currently am utilizing synoAI with deepstack (cpu version) on xpenology DSM 6.2 with great success. 

SynoAI info:
Docker installed (latest version) on DVA1622 7.1 - camera Trigger appears to communicate fine with synoai  

part of synoAI appsettings.json
"AI": {
    "Type": "CodeProjectAIServer",
    "Url": "http://192.168.42.249:32168"

SynoAI STdout (this output is being generated with a trigger "test" from the action rule I created)
2023-02-05T15:01:13.310832895Z,stdout,      cam1: Removing running camera block.
2023-02-05T15:01:13.310784600Z,stdout, [40m [37mdbug [39m [22m [49m: SynoAI.Controllers.CameraController[0]
2023-02-05T15:01:13.310733194Z,stdout,      cam1: Failed to get get predictions.
2023-02-05T15:01:13.310658677Z,stdout, [41m [30mfail [39m [22m [49m: SynoAI.Controllers.CameraController[0]
2023-02-05T15:01:13.310466468Z,stdout,      cam1: DeepStackAI: Failed with unknown error.
2023-02-05T15:01:13.310409073Z,stdout, [40m [1m [33mwarn [39m [22m [49m: SynoAI.Services.AIService[0]

**cam1: DeepStackAI: Responded with {"error":"The request timed out (#reqid {request.reqid})","success":false,"code":-1,"inferenceMs":0,"processMs":0,"analysisRoundTripMs":0}.**

2023-02-05T15:01:13.310109318Z,stdout, [40m [37mdbug [39m [22m [49m: SynoAI.Services.AIService[0]
2023-02-05T15:01:13.268952784Z,stdout,      cam1: DeepStackAI: POSTing image with minimum confidence of 0.45 (45%) to http://192.168.42.249:32168/v1/vision/detection.
2023-02-05T15:01:13.268789685Z,stdout, [40m [37mdbug [39m [22m [49m: SynoAI.Services.AIService[0]
2023-02-05T15:01:13.268691155Z,stdout,      cam1: Snapshot 1 of 1 received at EVENT TIME 110ms.
2023-02-05T15:01:13.268610449Z,stdout, [40m [32minfo [39m [22m [49m: SynoAI.Controllers.CameraController[0]
2023-02-05T15:01:13.268462784Z,stdout,      cam1: Snapshot received in 109ms.
2023-02-05T15:01:13.268344075Z,stdout, [40m [32minfo [39m [22m [49m: SynoAI.Controllers.CameraController[0]
2023-02-05T15:01:13.267785613Z,stdout,      cam1: Reading snapshot

CodeProjectAI Info:
Windows 10 VM created in Proxmox
Full install, utilizing GPU and CUDA (server is operating with no errors)
Installed modules : Objection Detection (YOLOv5 6.2) GPU (CUDA) enabled

Started CodeProjectAI via CMDline with the below VISION-DETECTION=True flag
https://static.codeproject.com/AI/docs/why/module_configuration.html
sc stop "CodeProject.AI Server"
sc start "CodeProject.AI Server" --VISION-DETECTION=True
Listening port: 32168? (to send images from synoai)
	
Active Connections from Windows OS hosting CodeProjectAI (not a Docker, windows 10)
  Proto  Local Address          Foreign Address        State
  TCP    0.0.0.0:135            CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:445            CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:3389           CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:5000           CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:5040           CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:32168          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49664          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49665          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49666          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49667          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49668          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49669          CodeProjectAI:0        LISTENING
  TCP    0.0.0.0:49725          CodeProjectAI:0        LISTENING
  TCP    127.0.0.1:32168        CodeProjectAI:49825    TIME_WAIT
  TCP    127.0.0.1:32168        CodeProjectAI:49835    TIME_WAIT
  TCP    192.168.42.249:139     CodeProjectAI:0        LISTENING
  TCP    [::]:135               CodeProjectAI:0        LISTENING
  TCP    [::]:445               CodeProjectAI:0        LISTENING
  TCP    [::]:3389              CodeProjectAI:0        LISTENING
  TCP    [::]:5000              CodeProjectAI:0        LISTENING
  TCP    [::]:32168             CodeProjectAI:0        LISTENING
  TCP    [::]:49664             CodeProjectAI:0        LISTENING
  TCP    [::]:49665             CodeProjectAI:0        LISTENING
  TCP    [::]:49666             CodeProjectAI:0        LISTENING
  TCP    [::]:49667             CodeProjectAI:0        LISTENING
  TCP    [::]:49668             CodeProjectAI:0        LISTENING
  TCP    [::]:49669             CodeProjectAI:0        LISTENING
  TCP    [::]:49725             CodeProjectAI:0        LISTENING
  TCP    [::1]:32168            CodeProjectAI:49824    TIME_WAIT
  TCP    [::1]:32168            CodeProjectAI:49826    TIME_WAIT
  TCP    [::1]:32168            CodeProjectAI:49828    TIME_WAIT
  TCP    [::1]:32168            CodeProjectAI:49834    TIME_WAIT
  TCP    [::1]:32168            CodeProjectAI:49836    TIME_WAIT
  TCP    [::1]:32168            CodeProjectAI:59966    TIME_WAIT
  UDP    0.0.0.0:123            *:*
  UDP    0.0.0.0:500            *:*
  UDP    0.0.0.0:3389           *:*
  UDP    0.0.0.0:4500           *:*
  UDP    0.0.0.0:5050           *:*
  UDP    0.0.0.0:5353           *:*
  UDP    0.0.0.0:5355           *:*
  UDP    127.0.0.1:1900         *:*
  UDP    127.0.0.1:56885        *:*
  UDP    127.0.0.1:56946        *:*
  UDP    192.168.42.249:137     *:*
  UDP    192.168.42.249:138     *:*
  UDP    192.168.42.249:1900    *:*
  UDP    192.168.42.249:56884   *:*
  UDP    [::]:123               *:*
  UDP    [::]:500               *:*
  UDP    [::]:3389              *:*
  UDP    [::]:4500              *:*
  UDP    [::]:5353              *:*
  UDP    [::]:5355              *:*
  UDP    [::1]:1900             *:*
  UDP    [::1]:56883            *:*
  
  Testing:
  Able to ping from synoAI to CodeProjectAI server (tested via bash under docker/synoai/terminal)
  Able to ping from CodeProjectAI to SynoAI docker container
  Fire wall disabled on both DSM hosting synoAI docker and CodeProjectAI Windows OS
  Attempted to use other ports within synoAI "Type": "CodeProjectAIServer", "Url": "http://192.168.42.249:83", 5000, 5040

**Comment by djdd87** [comment]
Can you provide your config, please?

**Comment by JoeyJoe6** [comment]
I just "##" out passwords, but kept it simple for testing. 

{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },

  "Url": "http://192.168.42.248:5000",
  "User": "synoai",
  "Password": "##",

  "MinSizeX": 100,
  "MinSizeY": 100,
  
  "DaysToKeepCaptures": 2,
  
  "AI": {
    "Type": "CodeProjectAIServer",
    "Url": "http://192.168.42.249:32168"
  },

  "Notifiers": [
    {
  "Type": "Email",
  "Sender": "event@internal.net",
  "Destination": "event@internal.net",
  "Host": "192.168.42.15",
  "Port": 25,
  "Username": "event",
  "Password": "##",
  "Encryption": "None"
    }
  ],

  "Cameras": [
    {
      "Name": "cam1",
      "Types": [ "Person, bicycle, car, motorbike, bus, truck, boat, cat, dog, horse" ], 
      "Threshold": 45,
      "MinSizeX": 150,
      "MinSizeY": 150
    },
  
	]
}

SynoAI log when test trigger sent:
      cam1: The camera is currently running; no other camera requests will be processed while this request is ongoing.
info: SynoAI.Controllers.CameraController[0]
      Captures Clean Up: Cleaning up images older than 2 day(s).
info: SynoAI.Controllers.CameraController[0]
      cam1: Snapshot 1 of 1 requested at EVENT TIME 0ms.
dbug: SynoAI.Services.SynologyService[0]
      cam1: Found with Synology ID '1'.
dbug: SynoAI.Services.SynologyService[0]
      cam1: Taking snapshot from 'webapi/entry.cgi?version=9&id=1&api="SYNO.SurveillanceStation.Camera"&method=GetSnapshot&profileType=1'.
info: SynoAI.Services.SynologyService[0]
      cam1: Taking snapshot
dbug: SynoAI.Services.SynologyService[0]
      cam1: Reading snapshot
info: SynoAI.Controllers.CameraController[0]
      cam1: Snapshot received in 92ms.
info: SynoAI.Controllers.CameraController[0]
      cam1: Snapshot 1 of 1 received at EVENT TIME 93ms.
dbug: SynoAI.Services.AIService[0]
      cam1: DeepStackAI: POSTing image with minimum confidence of 0.45 (45%) to http://192.168.42.249:32168/v1/vision/detection.
dbug: SynoAI.Services.AIService[0]
      cam1: DeepStackAI: Responded with {"success":false,"error":"Error occured on the server","code":500,"analysisRoundTripMs":109}.
warn: SynoAI.Services.AIService[0]
      cam1: DeepStackAI: Failed with unknown error.
fail: SynoAI.Controllers.CameraController[0]
      cam1: Failed to get get predictions.
dbug: SynoAI.Controllers.CameraController[0]
      cam1: Removing running camera block.

codeproject syslog:
09:43:27:Client request 'detect' in the queue (...5d5443)
09:43:27:Request 'detect' dequeued for processing (...5d5443)
09:43:27:Object Detection (YOLOv5 6.2): Retrieved objectdetection_queue command
09:43:27:Response received (...5d5443)
09:43:27:Object Detection (YOLOv5 6.2):  [Exception] : Traceback (most recent call last):
09:43:27:Object Detection (YOLOv5 6.2): Queue and Processing Object Detection (YOLOv5 6.2) command 'detect' (...5d5443) took 16ms

**Seems to be an instant failure on SynoAI's end. Is there a timeout value I can modify to wait for a response longer?

**Reply by djdd87** [reply]
Try and define your types properly:

```json
"Types": [ "Person", "bicycle", "car", "motorbike", "bus", "truck", "boat", "cat", "dog", "horse" ],
```

**Reply by JoeyJoe6** [reply]
Thanks for the eyes. Modified the types, but this did not help.
**Final Solution**: SynoAI did not like working with SS CMS. I noticed that I forget to mention I was using CMS to control my DMS's. I installed SynoAI on each of the SS setups and modified appsettings for each instance. ALL IS WORKING NOW :-).. THANK YOU for a great product!!

**Reply by djdd87** [reply]
I've no idea what that even is! So well done for sorting it.

**Comment by JoeyJoe6** [comment]
I also see this popup on DMS7 interface. 
![image](https://user-images.githubusercontent.com/116003744/217374587-bc1e5af0-9e26-41a9-bcb1-e3ce45311fd0.png)
